File: 백준/Gold/9370. 미확인 도착지/미확인 도착지.py
from heapq import heappop, heappush
import sys
input = sys.stdin.readline
INF = float('inf')

def dijkstra(st: int):
    hq = []

    distance = [INF] * (n + 1)
    distance[st] = 0
    heappush(hq, (distance[st], False, st))
    used = [False] * (n + 1)

    while hq:
        dist, vis, cur = heappop(hq)
        if distance[cur] < dist: continue
        for nxt,cost in graph[cur]:
            tmp = vis

            if (cur,nxt) == (g,h) or (nxt,cur) == (g,h):
                tmp = True

            if dist+cost < distance[nxt]:
                distance[nxt] = dist + cost
                used[nxt] = tmp
                heappush(hq, (distance[nxt], tmp, nxt))
            elif not used[nxt] and tmp and cost+dist==distance[nxt]:
                used[nxt] = tmp
                heappush(hq, (distance[nxt], tmp, nxt))
            
    return sorted([i for i in dest if used[i]])

t = int(input())
for _ in range(t):
    n,m,k = map(int,input().split())
    s,g,h = map(int,input().split())

    graph = [[] for _ in range(n + 1)]
    for _ in range(m):
        u,v,w = map(int,input().split())
        graph[u].append((v,w))
        graph[v].append((u,w))
    dest = [int(input()) for _ in range(k)]

    dist = dijkstra(s)
    print(*dist)

File: 백준/Gold/9370. 미확인 도착지/미확인 도착지.rs
use std::cmp::Reverse;
use std::collections::BinaryHeap;
use std::io::{read_to_string, stdin};

fn dijkstra(st: usize, n: usize, g: usize, h: usize, graph: &Vec<Vec<(usize, u32)>>, dest: &Vec<usize>) -> Vec<usize> {
    let mut hq = BinaryHeap::new();
    let mut distance = vec![u32::MAX; n + 1];
    distance[st] = 0;
    hq.push((Reverse(distance[st]), false, st));
    let mut used = vec![false; n + 1];

    while let Some((Reverse(dist), vis, cur)) = hq.pop() {
        if distance[cur] < dist {
            continue;
        }
        for &(nxt, cost) in &graph[cur] {
            let mut tmp = vis;
            if (cur == g && nxt == h) || (cur == h && nxt == g) {
                tmp = true;
            }
            if dist + cost < distance[nxt] {
                distance[nxt] = dist + cost;
                used[nxt] = tmp;
                hq.push((Reverse(distance[nxt]), tmp, nxt));
            } else if !used[nxt] && tmp && dist + cost == distance[nxt] {
                used[nxt] = tmp;
                hq.push((Reverse(distance[nxt]), tmp, nxt));
            }
        }
    }

    let mut res: Vec<usize> = dest.iter().cloned().filter(|&i| used[i]).collect();
    res.sort();
    res
}

fn main() {
    let stdin = read_to_string(stdin()).unwrap();
    let mut tokens = stdin.split_whitespace();
    let mut next = || tokens.next().unwrap();
    let mut res = String::new();

    let t: usize = next().parse().unwrap();
    for _ in 0..t {
        let (n, m, k): (usize, usize, usize) = (next().parse().unwrap(), next().parse().unwrap(), next().parse().unwrap());
        let (s, g, h): (usize, usize, usize) = (next().parse().unwrap(), next().parse().unwrap(), next().parse().unwrap());
        let mut graph = vec![Vec::new(); n + 1];
        for _ in 0..m {
            let (u, v, w): (usize, usize, u32) = (next().parse().unwrap(), next().parse().unwrap(), next().parse().unwrap());
            graph[u].push((v, w));
            graph[v].push((u, w));
        }

        let mut dest = Vec::with_capacity(k);
        for _ in 0..k {
            let d: usize = next().parse().unwrap();
            dest.push(d);
        }

        let dist = dijkstra(s, n, g, h, &graph, &dest);
        for d in dist {
            res.push_str(d.to_string().as_str());
            res.push(' ');
        }
        res.push('\n');
    }

    print!("{res}")
}
